Addresses #4362, add support of HeatPump:PlantLoop:EIR:Heating to model air-to-water heat pump

joseph-robertson opened the pull request:
Pull request overview
---------------------

Tests demonstrating https://github.com/NREL/OpenStudio/pull/4368.

TODO:
- [x] AirSource
- [x] WaterSource

Link to relevant GitHub Issue(s) if appropriate:

Link to the **Linux.deb** installer to use for CI Testing. If not set, it will default to latest official release.
[OpenStudio Installer]: http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/incremental/develop/4368/OpenStudio-3.2.2-alpha%2B2f7221d03b-Ubuntu-18.04.deb

This Pull Request is concerning:

 - [x] **Case 1 - `NewTest`:** a new test for a new model API class,
 - [ ] **Case 2 - `TestFix`:** a fix for an existing test. The GitHub issue should be referenced in the PR description
 - [ ] **Case 3 - `NewTestForExisting`:** a new test for an already-existing model API class
 - [ ] **Case 4 - `Other`:** Something else, like maintenance of the repo, or just committing test results with a new OpenStudio version.

Depending on your answer, please fill out the required section below, and delete the three others.
Leave the review checklist in place.


----------------------------------------------------------------------------------------------------------

### Case 1: New test for a new model API class

Please include which class(es) you are adding a test to specifically test for.
Include a link to the OpenStudio Pull Request in which you are adding the new classes, or the class itself if already on develop.
<!---
> eg:
>
> This pull request is in relation with the Pull Request [NREL/OpenStudio#3031](https://github.com/NREL/OpenStudio/pull/3031), and  will specifically test for the following classes:
> * `AirTerminalSingleDuctConstantVolumeFourPipeBeam`
> * `CoilCoolingFourPipeBeam`
> * `CoilHeatingFourPipeBeam`
> * Additionally it explicitly tests for the existing class [TableMultiVariableLookUp](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/TableMultiVariableLookup.hpp)
-->

#### Work Checklist

The following has been checked to ensure compliance with the guidelines:

 - [ ] Tests pass either:
     - [ ] with official OpenStudio release (include version):
         - [ ] A matching OSM test has been added from the successful run of the Ruby one with the official OpenStudio release
         - [ ] The label `AddedOSM` has been added to this PR
         - [ ] All new `out.osw` have been committed

     - [ ] with current develop (incude SHA):
         - [ ] The label `PendingOSM` has been added to this PR
         - [ ] A matching OSM test has not yet been added because the official release is pending, but `model_tests.rb` has a TODO.
            ```ruby
            def test_airterminal_cooledbeam_rb
              result = sim_test('airterminal_cooledbeam.rb')
            end

            # TODO: To be added in the next official release after: 2.5.0
            # def test_airterminal_fourpipebeam_osm
            #   result = sim_test('airterminal_fourpipebeam.osm')
            # end
            ```
        - [ ] No `out.osw` have been committed as they need to be run with an official OpenStudio version


 - [ ] **Ruby test is stable**: when run multiple times on the same machine, it produces the same total site kBTU.
    Please paste the heatmap png generated after running the following commands:
     - [ ] I ensured that I assign systems/loads/etc in a repeatable manner (eg: if I assign stuff to thermalZones, I do `model.getThermalZones.sort_by{|z| z.name.to_s}.each do ...` so I am sure I put the same ZoneHVAC systems to the same zones regardless of their order)
     - [ ] I tested stability using `process_results.py` (see `python process_results.py --help` for usage).
    Please paste the text output or heatmap png generated after running the following commands:
        ```bash
        # Clean up all custom-tagged OSWs
        python process_results.py test-stability clean
        # Run your test 5 times in a row. Replace `testname_rb` (eg `airterminal_fourpipebeam_rb`)
        python process_results.py test-stability run -n testname_rb
        # Check that they all passed
        python process_results.py test-status --tagged
        # Check site kBTU differences
        python process_results.py heatmap --tagged

        ```

 - [ ] **Object has been added to `autosize_hvac.rb` to ensure the autosizedXXX values methods do work**


----------------------------------------------------------------------------------------------------------

### Review Checklist

 - [ ] Code style (indentation, variable names, strip trailing spaces)
 - [ ] Functional code review (it has to work!)
 - [ ] Matching OSM test has been added or `# TODO` added to `model_tests.rb`
 - [ ] Appropriate `out.osw` have been committed
 - [ ] Test is stable
 - [ ] Object is tested in `autosize_hvac` as appropriate
 - [ ] The appropriate labels have been added to this PR:
   - [ ] One of: `NewTest`, `TestFix`, `NewTestForExisting`, `Other`
   - [ ] If `NewTest`: add `PendingOSM` or `AddedOSM`

joseph-robertson committed:
Stub test for heatpump plantloop.

joseph-robertson committed:
Add new test to model tests file.

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-linux], Rubocop 0.81.0)

joseph-robertson committed:
Name the curves.

joseph-robertson committed:
Merge branch 'issue-4362' of github.com:NREL/OpenStudio-resources into issue-4362

joseph-robertson committed:
Add pumps and stpt managers.

jmarrec commented on model_tests.rb:
```diff
@@ -530,6 +530,15 @@ def test_heatpump_hot_water_osm
+    result = sim_test('heatpump_plantloop_eir.rb')
+  end
+
+  # TODO: To be added in the next official release after: 3.2.0
```
```suggestion
  # TODO: To be added in the next official release after: 3.2.1
```

jmarrec commented on model/simulationtests/heatpump_plantloop_eir.rb:
```diff
@@ -0,0 +1,149 @@
+hw_pump.setPumpControlType('Intermittent')
+hw_pump.addToNode(hw_loop.supplyInletNode)
+
+cw_loop = OpenStudio::Model::PlantLoop.new(model)
```
I'd move this loop either above or below the two other. perhaps make it into a block comment

```
############   C O N D E N S E R    S I D E  ##################
cw_loop = ...

############ HEATING / COOLING (LOAD) LOOPS  ############
hw_loop = ...

chw_loop = ...
```

jmarrec replied:
Also, from a cursory look, you don't have an equipment on the supply side of this condenser loop do you? @joseph-robertson

jmarrec commented:
@joseph-robertson I don't expect this test to work yet without the changes to demand/supply inlet ports I asked on the SDK side.

Note that you'll also need to modify the autosize test to add this object before the test PR changes workflow passes.

joseph-robertson committed:
Get test simulating.

joseph-robertson committed:
Add AirSource scenario.

jmarrec committed:
Adjust test to use a GroundHX on the condenser loop, and connect the HW/CHW loops to a FourPipeFanCoil. Adjust the airsource/watersource now that it's enforced in model. IDF is valid, but it crashes

```
   ************* Beginning Simulation
   ** Severe  ** PsyRhoAirFnPbTdbW: RhoAir (Density of Air) is calculated <= 0 [-1.42729E-002].
   **   ~~~   ** pb =[99125.00], tdb=[-24471.37], w=[0.0000000].
   **   ~~~   **  Routine=Unknown, Environment=RUN PERIOD 1, at Simulation time=01/01 00:00 - 00:15
   **  Fatal  ** Program terminates due to preceding condition.
   ...Summary of Errors that led to program termination:
   ..... Reference severe error count=1
```

jmarrec committed:
Use the right SourceSideFlow rates for the AirSource ones

jmarrec committed:
rename PLHPs for clarity

jmarrec committed:
Merge branch 'develop' into issue-4362

jmarrec committed:
Prepare for autosize_hvac (but it's failing, while it works on develop) + rubocop

The addition of the plhp stuff is making E+ crash... haven't managed to fix it...

jmarrec committed:
Try something else, still failing!

jmarrec committed:
Keep trying, still failing!

jmarrec committed:
NOPEEEE

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux], Rubocop 0.81.0)

jmarrec committed:
Made a typo...

jmarrec committed:
Try to disable the HeatPump:PlantLoop:EIR in autosize_hvac test completely

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux], Rubocop 0.81.0)

jmarrec committed:
Merge branch 'develop' into issue-4362

jmarrec committed:
Apparently even if I harsize the Reference Capacity, it's still reported in the SQL... I had to add aliases too

@joseph-robertson FYI, by changing the name of the methods instead of matching the IDD, aliases are needed here for the autosizedXXX stuff. 
see https://github.com/NREL/OpenStudio/pull/4368#pullrequestreview-723142659 for more info

jmarrec committed:
Update test files following the rename of the methods

https://github.com/NREL/OpenStudio/commit/3d1f5976b561c2f119d956313d9e333413beda74

jmarrec commented:
Re running after getting latest installer http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/incremental/develop/4368/OpenStudio-3.2.2-alpha%2B2f7221d03b-Ubuntu-18.04.deb
